wolfcrypt/src/port/caam/caam_qnx.c

```diff
@@ -0,0 +1,1262 @@
+ *
+ * wolfSSL is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+#include "caam_driver.h"
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <pthread.h>
+#include <sys/iofunc.h>
+#include <sys/dispatch.h>
+#include <sys/neutrino.h>
+#include <sys/resmgr.h>
+#include <devctl.h>
+
+/* virtual address for accessing CAAM addresses */
+uintptr_t virtual_base = 0;
+
+/* keep track of which ID memory belongs to so it can be free'd up */
+#define MAX_PART 7
+pthread_mutex_t sm_mutex;
+int sm_ownerId[MAX_PART];
+
+/* variables for I/O of resource manager */
+resmgr_connect_funcs_t connect_funcs;
+resmgr_io_funcs_t      io_funcs;
+dispatch_t             *dpp;
+resmgr_attr_t          rattr;
+dispatch_context_t     *ctp;
+iofunc_attr_t          ioattr;
+
+int io_devctl (resmgr_context_t *ctp, io_devctl_t *msg, iofunc_ocb_t *ocb);
+int io_open (resmgr_context_t *ctp, io_open_t  *msg, RESMGR_HANDLE_T *handle,
+        void *extra);
+int io_read (resmgr_context_t *ctp, io_read_t  *msg, RESMGR_OCB_T *ocb);
+int io_write(resmgr_context_t *ctp, io_write_t *msg, RESMGR_OCB_T *ocb);
+int io_close_ocb(resmgr_context_t *ctp, void *reserved, RESMGR_OCB_T *ocb);
+
+
+/* read the contents at offset from BASE address */
+unsigned int CAAM_READ(unsigned int ofst) {
+    return in32(virtual_base + ofst);
+}
+
+
+/* takes in offset from BASE address */
+void CAAM_WRITE(unsigned int ofst, unsigned int in)
+{
+    out32(virtual_base + ofst, in);
+}
+
+
+/* Sets the base address to use for read/write
+ * returns 0 on success
+ */
+int CAAM_SET_BASEADDR()
+{
+    /* address range for CAAM is CAAM_BASE plus 0x10000 */
+    virtual_base = mmap_device_io(0x00010000, CAAM_BASE);
+    if (virtual_base == (uintptr_t)MAP_FAILED) {
+        WOLFSSL_MSG("Unable to map virtual memory");
+        return -1;
+    }
+    return 0;
+}
+
+
+
+/* convert a virtual address to a physical address
+ * returns the physical address on success
+ */
+unsigned int CAAM_ADR_TO_PHYSICAL(void* in, int inSz)
+{
+    off64_t ofst = 0;
+    int ret, count = 0;;
+
+    if (in == NULL)
+        return 0;
+
+    if (inSz == 0)
+        inSz = 1;
+
+    do {
+        ret = mem_offset64(in, NOFD, inSz, &ofst, NULL);
+        if (ret != 0) {
+            WOLFSSL_MSG("posix offset failed");
+        #if defined(WOLFSSL_CAAM_DEBUG) || defined(WOLFSSL_CAAM_PRINT)
+            perror("");
+        #endif
+        }
+        msync(in, inSz, MS_INVALIDATE);
+        count++;
+    } while (ret != 0 && ret == -1 && count < 5);
+
+    return (int)ofst;
```

Comment:
Is the physical address address guaranteed to fit in a signed int?  That is, is an off 32/64 _t type more robust here?

Reply:
With the IMX6UL all address pointers that can be used by the CAAM are 32 bits. That's not to say though that there could be a potential for larger addresses on a different device. I'll switch it to be CAAM_ADDRESS return type which can be set at build time.